controller/gke-cluster-config-handler.go

```diff
@@ -0,0 +1,384 @@
+	if config.Spec.Imported {
+		logrus.Infof("cluster [%s] is imported, will not delete GKE cluster", config.Name)
+		return config, nil
+	}
+	if config.Status.Phase == gkeConfigNotCreatedPhase {
+		// The most likely context here is that the cluster already existed in GKE, so we shouldn't delete it
+		logrus.Warnf("cluster [%s] never advanced to creating status, will not delete GKE cluster", config.Name)
+		return config, nil
+	}
+
+	logrus.Infof("deleting cluster [%s]", config.Name)
+
+	//TODO: delete cluster
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	svc, err := utils.GetServiceClient(ctx, config.Spec.CredentialContent)
+	if err != nil {
+		return config, err
+	}
+
+	logrus.Debugf("Removing cluster %v from project %v, region/zone %v", config.Spec.ClusterName, config.Spec.ProjectID, config.Spec.Region)
+	operation, err := utils.WaitClusterRemoveExp(ctx, svc, config)
+
+	if err != nil && !strings.Contains(err.Error(), "notFound") {
+		return config, err
+	} else if err == nil {
+		logrus.Debugf("Cluster %v delete is called. Status Code %v", config.Spec.ClusterName, operation.HTTPStatusCode)
+	} else {
+		logrus.Debugf("Cluster %s doesn't exist", config.Spec.ClusterName)
+	}
+
+	return config, nil
+}
+
+func (h *Handler) create(config *gkev1.GKEClusterConfig) (*gkev1.GKEClusterConfig, error) {
+	/*
+		if err := utils.ValidateCreateRequest(config); err != nil {
+			return config, err
+		}
+	*/
+	if config.Spec.Imported {
+		config = config.DeepCopy()
+		config.Status.Phase = gkeConfigImportingPhase
+		return h.gkeCC.UpdateStatus(config)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	svc, err := utils.GetServiceClient(ctx, config.Spec.CredentialContent)
+	if err != nil {
+		return config, err
+	}
+
+	createClusterRequest, err := utils.GenerateGkeClusterCreateRequest(config)
+	if err != nil {
+		return config, err
+	}
+
+	operation, err := svc.Projects.Locations.Clusters.Create(
+		utils.LocationRRN(config.Spec.ProjectID, config.Spec.Region),
+		createClusterRequest).Context(ctx).Do()
+
+	if err != nil && !strings.Contains(err.Error(), "alreadyExists") {
+		return config, err
+	}
+	if err == nil {
+		logrus.Debugf("Cluster %s create is called for project %s and region/zone %s. Status Code %v",
+			config.Spec.ClusterName, config.Spec.ProjectID, config.Spec.Region, operation.HTTPStatusCode)
+	}
+
+	config = config.DeepCopy()
+	config.Status.Phase = gkeConfigCreatingPhase
+	return h.gkeCC.UpdateStatus(config)
+}
+
+func (h *Handler) checkAndUpdate(config *gkev1.GKEClusterConfig) (*gkev1.GKEClusterConfig, error) {
+
+	if err := h.validateUpdate(config); err != nil {
+		config = config.DeepCopy()
+		config.Status.Phase = gkeConfigUpdatingPhase
+		var updateErr error
+		config, updateErr = h.gkeCC.UpdateStatus(config)
+		if updateErr != nil {
+			return config, updateErr
+		}
+		return config, err
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
```

Comment:
Wondering how often this could save us, considering using it else where too. Figured since we aren't passing down the parent context there wasn't much we could use the param for without making adjustments.